testing-plugin/sources/generic/generic-grammar.ym

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2022 Shikhar Vashistha
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+%code top {
+#include "generic-parser.h"
+
+}
+
+
+%code {
+
+#include "cfg-grammar-internal.h"
+#include "cfg-parser.h"
+#include "syslog-names.h"
+#include "messages.h"
+#include "plugin.h"
+
+#include <string.h>
+
+}
+
+%define api.prefix {generic_}
+
+/* this parameter is needed in order to instruct bison to use a complete
+ * argument list for yylex/yyerror */
+
+%lex-param {CfgLexer *lexer}
+%parse-param {CfgLexer *lexer}
+%parse-param {LogDriver **instance}
+%parse-param {gpointer arg}
+
+/* INCLUDE_DECLS */
+
+%%
+
+start
+  : LL_CONTEXT_INNER_DEST first_rule          { YYACCEPT; }
```

Comment:
This file defines the configuration language of the plugin.
It is a Bison grammar file, you can read about it more here:
https://www.gnu.org/software/bison/manual/html_node/Grammar-File.html

If you, for example, want a plugin called `file-rotation()` with options like `size(100MB)` or `interval("daily")`, you can define something like this:

```
start
	: LL_CONTEXT_INNER_DEST file_rotation { YYACCEPT; }
	;

file_rotation
        : KW_FILE_ROTATION
          {
            last_plugin = file_rotation_new();
            *instance = (LogDriverPlugin*)last_plugin;
          }
          '(' file_rotation_options ')'
        ;

file_rotation_options
        : file_rotation_option file_rotation_options
        |
        ;

file_rotation_option
        : KW_SIZE '(' nonnegative_integer ')'       { file_rotation_set_size(last_driver, $3); }
        | KW_INTERVAL '(' string ')'                          { log_rotation_set_interval(last_driver, $3); free($3); }
        ;
```

And then, because you defined this plugin as an INNER_DEST plugin, you can use it inside the file destination, for example:

```
destination {
  file("/var/log/messages" file-rotation(size(100MB), interval("daily")));
};
```